docs/coding-style.md

```diff
@@ -0,0 +1,325 @@
+called `git_repository_open()`. Functions that are not public but declared in
+an internal header file for use by other subsystems should follow the same
+naming pattern. File-local static functions must not have a `git` prefix, but
+should have a prefix indicating their respective subsystem.
+
+All structures declared in the libgit2 project must have a `typedef`, we do not
+use `struct type` variables. Type names follow the same schema as functions.
+
+### Error Handling
+
+The libgit2 project mostly uses error codes to indicate errors. Error codes are
+always of type `int`, where `0` indicates success and a negative error code
+indicates an error case. In some cases, positive error codes may be used to
+indicate special cases. Returned values that are not an error code should be
+returned via an out value:
+
+```c
+int doit(const char **out, int arg)
+{
+	if (!arg)
+		return -1;
+	*out = "Got an argument";
+	return 0;
+}
+```
+
+To avoid repetitive and fragile error handling in case a function has resources
+that need to be free'd, we use `goto out`s:
+
+```c
+int doit(char **out, int arg)
+{
+	int error = 0;
+	char *c;
+
+	c = malloc(strlen("Got an argument") + 1);
+	if (!c) {
+		error = -1;
+		goto out;
+	}
+
+	if (!arg) {
+		error = -1;
+		goto out;
+	}
+
+	strcpy(c, "Got an argument")
+	*out = c;
+
+out:
+	if (error)
+		free(c);
+	return error;
+}
+```
+
+When calling functions that return an error code, you should assign the error
+code to an `error` variable and, in case an error case is indicated and no
+custom error handling is required, return that error code:
+
+```c
+int foobar(void)
+{
+	int error;
+
+	if ((error = doit()) < 0)
+		return error;
+
+	return 0;
+}
+```
+
+When doing multiple function calls where all of the functions return an error
+code, it's common practice to chain these calls together:
+
+```c
+int doit(void)
+{
+	int error;
+
+	if ((error = dothis()) < 0 ||
+	    (error = dothat()) < 0)
+		return error;
+
+	return 0;
+}
+```
+
+## CMake Coding Style
+
+The following section defines the coding style for our CMake build system.
+
+### Indentation
+
+Code is indented by tabs, where a tab is 8 spaces. Each opening scope increases
+the indentation level.
+
+```cmake
+IF(CONDITION)
+	DOIT()
+ENDIF()
+```
+
+### Spaces
+
+There must be no space between keywords and their opening brace. While this is
+the same as in our C codebase for function calls, this also applies to
+conditional keywords. This is done to avoid the awkward-looking `ELSE ()`
+statement.
+
+```cmake
+IF(CONDITION)
+	DOIT()
+ELSE()
+	DOTHAT()
+ENDIF()
+```
+
+### Naming
+
+While CMake is completely case-insensitive when it comes to function calls, we
+want to agree on a common coding style for this. To reduce the danger of
+repetitive strain injuries, all function calls should be lower-case (NB: this is
+not currently the case yet, but introduced as a new coding style by this
+document). Similarly, declarations of functions should be lower-case, too.
+
```

Comment:
woof, I don't know how I feel about this.  I don't _love_ that our CMake config is mostly ALL UPPERCASE, it does feel a little more "right" to me to have it that way instead of a mix of ALL UPPER for variables and all_lower for functions.

Reply:
Haha, I knew this is going to be a tough sell ;) Thing is, CMake _always_ uses UPPERVARS and lowerfuncs. Take a look at their documentation, their exposed variables, "An Introduction to Modern CMake" (https://cliutils.gitlab.io/modern-cmake/). I feel like we shouldn't try to be special here and just do what's common practice.

Reply:
And by the way, I only mean this repetitive strain remark half-jokingly: I always got a quite noticeable pain in my little fingers after typing CMake code for more than half an hour :(

Reply:
From their documentation, it looks like they also lower-case things like `if`, `else`...  should we adopt that, too?